samza-rest/src/main/java/org/apache/samza/rest/proxy/job/YarnRestJobStatusProvider.java

```diff
@@ -0,0 +1,164 @@
+
+  @Override
+  public void getJobStatuses(Collection<Job> jobs) throws IOException, InterruptedException {
+    if (jobs == null || jobs.isEmpty()) {
+      return;
+    }
+
+    // We will identify jobs returned by the YARN application states by their qualified names, so build a map
+    // to translate back from that name to the JobInfo we wish to populate. This avoids parsing/delimiter issues.
+    final Map<String, Job> qualifiedJobToInfo = new HashMap<>();
+    for(Job job : jobs) {
+      qualifiedJobToInfo.put(getQualifiedJobName(new JobInstance(job.getJobName(), job.getJobId())), job);
+    }
+
+    // Hit the rest endpoint and fetch the current status
+    try {
+      byte[] response = httpGet(_apiEndpoint);
+      ApplicationInfoWrapper appInfoWrapper = _objectMapper.readValue(response, ApplicationInfoWrapper.class);
+      for (ApplicationInfo appInfo:  appInfoWrapper.getApps()) {
+        String qualifiedName = appInfo.getName();
+        JobStatus samzaStatus = yarnStateToSamzaStatus(YarnApplicationState.valueOf(appInfo.getState().toUpperCase()));
+        Job job = qualifiedJobToInfo.get(qualifiedName);
+
+        // If job is null, it wasn't requested.  The default status is STOPPED because there could be many
+        // application attempts in that status. Only update the job status if it's not STOPPED.
+        if (job != null && (job.getStatusDetail() == null || samzaStatus != JobStatus.STOPPED)) {
+          job.setStatusDetail(appInfo.getState());
+          job.setStatus(samzaStatus);
+        }
+      }
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to retrieve node info.", e);
+    }
+  }
+
+  @Override
+  public Job getJobStatus(JobInstance jobInstance)
+      throws IOException, InterruptedException {
+    Job info = new Job(jobInstance.getJobName(), jobInstance.getJobId());
+    getJobStatuses(Collections.singletonList(info));
+    return info;
+  }
+
+  /**
+   * Constructs the job name used in YARN. This is the value returned by the "name"
+   * attribute form the Resource Manager API /ws/v1/cluster/apps.
+   *
+   * @param jobInstance the instance of the job.
+   * @return            the job name to use for the job in YARN.
+   */
+  public static String getQualifiedJobName(JobInstance jobInstance) {
+    return String.format(JOB_NAME_ID_FORMAT, jobInstance.getJobName(), jobInstance.getJobId());
+  }
+
+  /**
+   * Translates the YARN application state to the more generic Samza job status.
+   *
+   * @param yarnState the YARN application state to translate.
+   * @return          the corresponding Samza job status.
+   */
+  private JobStatus yarnStateToSamzaStatus(YarnApplicationState yarnState) {
+    switch (yarnState) {
+      case RUNNING:
+        return JobStatus.STARTED;
+      case NEW:
+      case NEW_SAVING:
+      case SUBMITTED:
+      case ACCEPTED:
+        return JobStatus.STARTING;
+      case FINISHED:
+      case FAILED:
```

Comment:
We don't have a FAILED job status?

Reply:
The REST API provides both the simplified JobStatus and the cluster specific status detail. The simplified status is primarily focused on whether or not the job is running or not. It's not really used for diagnosing problems. 

This area of the code was actually my first contribution to Samza, and I have to admit, it's a bit naive in retrospect. But this patch is mostly about producing equivalent behavior with better performance, so I wouldn't recommend changing it at the moment.

Reply:
Makes sense, we can always add the failed status if we need it.